### ISSUE TITLE
Domain Expiry Monitor

### ISSUE THREAD

**Issue by shmaltz** [louislam/uptime-kuma#944]
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

New Monitor

### 🔖 Feature description

It would be awesome if we can add domains, and monitor their expiry status.

### ✔️ Solution

We should be notified when there is a change to the DNS or expiry date, and be notified x days before expiry.

### 📝 Additional Context

11Uptime has a similar feature, as well as Better Uptime

**Comment by Rmlyy** [comment]
That would be awesome!

**Comment by BCsabaEngine** [comment]
I think it is hard to retrieve domain information from tld companies/organization. For example there is no public api for hungarian domains.

If you want to detect dns (host->ip) resolve changes that is other issue.

**Comment by pepa65** [comment]
> it is hard to retrieve domain information from tld companies/organization

It is hard and not reliable without a standardized API, but for many domains WHOIS scraping would go a long way. Still not too fun to implement...

**Comment by shmaltz** [comment]
Seems like there are quite a few APIs, some with big limits, and some completely free:
https://blog.api.rakuten.net/top-domain-whois-api/

**Comment by crbon** [comment]
This would be awesome +1 🎉

**Comment by louislam** [comment]
Just found a library. I think it is quite promising, because it directly queries the whois 43 port server.
As long as Uptime Kuma caches the expiration date, it should not hit the rate limit.
https://github.com/eithan1231/whois-light

**Comment by BeecherNetworks** [comment]
+1 for this. I was looking for something self-hosted for this when I came across domain-monitor.io, and that looked suspiciously like Uptime Kuma so I came over here to search. This and [SSL expiry notifications](https://github.com/louislam/uptime-kuma/issues/573) would be super cool.

**Comment by karelkryda** [comment]
> Just found a library. I think it is quite promising, because it directly queries the whois 43 port server. As long as Uptime Kuma caches the expiration date, it should not hit the rate limit. https://github.com/eithan1231/whois-light

According to open issues, it seems that the library is not completely consistent for all TLDs.
This could have a negative effect on the implementation into Uptime Kuma, as it would not be possible to have one code for all TLDs.

If you are interested, I could try to look at it and try to make a universal WHOIS library for most domains. But I can't promise anything.

Would that help?

**Comment by karelkryda** [comment]
@louislam As I wrote in the comment above, I started working on a library to retrieve domain data from WHOIS servers.

I'm posting a [link](https://github.com/karelkryda/whois) to my GitHub repository.

**Comment by kotBanditsky** [comment]
You're overcomplicating it, just let users enter an expiration date for the domain name when adding a new monitor. And then send notifications when the specified date approaches.

**Comment by anseltaft** [comment]
+1 for this. I'm commenting in support because I can't tell where things landed with the request, though I did see the Github repo and closed, duplicate request.

Regarding the comment before mine, I have to say it completely misses the point: the reason for the feature request is to account for absentmindedness/human error and if we all took the time to track when domains were due for renewal then there would be no need for such a thing... but people are going to people. I love the idea of Uptime Kuma offering an additional check, and I already have every domain set to autorenew. Sometimes things happen and people forget to update their credit card's expiration date for instance.

**Comment by karelkryda** [comment]
@anseltaft In fact, a big problem is the inconsistent WHOIS servers of different companies. As I wrote above, I started working on the Universal WHOIS library, but basically most registrars had different results when requesting information about the domain. Personally, I think that we are currently unable to provide full support for most domain extensions, and therefore this requirement is in a frozen state. 

I'm not sure if I once posted recent changes to GH, but if you look at my repository, you'll probably understand what I'm talking about.

**Comment by alwynispat** [comment]
Go with the manual approach first, then work on the automation later.

**Comment by karelkryda** [comment]
@alexandernst In fact, I thought about it too. I planned to eventually use it in my library, and it could then be used as the only dependency. But I haven't gotten to it yet... maybe on the weekend.
However, it is true that at the moment this is probably the only solution.

**Comment by nickistre** [comment]
Would love to have this feature, even if it's something manually entered somewhere for now (Having "Check domain expiry for xxx.tld" show up in slack at 30, 14, and 7 days before would be just fine to let the team know to check in this, and update the date in Uptime Kuma if needed).

On a side note, I did discover this check domain expiry plugin for Nagios, and it's not trivial (it's written in SH, and uses the `whois` tool directly). Search for the `get_expiration()` function.

https://github.com/glensc/monitoring-plugin-check_domain/blob/master/check_domain.sh

**Comment by nogweii** [comment]
`RDAP` is the replacement for WHOIS, see some more details [on Wikipedia](https://en.wikipedia.org/wiki/Registration_Data_Access_Protocol). It's very useful as it implements a structured JSON format for representing domain metadata, including the expiration.

The way it'd work is that Uptime-Kuma would have to cache the RDAP Bootstrap list (`https://data.iana.org/rdap/dns.json`) to get a list of TLD -> RDAP servers and then query the RDAP server on a regular basis for the expiry date.

**Comment by louislam** [comment]
> `RDAP` is the replacement for WHOIS, see some more details [on Wikipedia](https://en.wikipedia.org/wiki/Registration_Data_Access_Protocol). It's very useful as it implements a structured JSON format for representing domain metadata, including the expiration.
> 
> The way it'd work is that Uptime-Kuma would have to cache the RDAP Bootstrap list (`https://data.iana.org/rdap/dns.json`) to get a list of TLD -> RDAP servers and then query the RDAP server on a regular basis for the expiry date.

Wow, it could be so helpful.

Surprisingly, I don't know why I didn't find anything about RDAP.

Haven't tested it yet, but I just found a nodejs package for rdap:
https://github.com/dcoffey-zengenti/node-rdap

**Comment by nickistre** [comment]
Looks like not all TLDs have working RDAP yet (`.io` seems to be one, see: https://openrdap.org/demo?cmd=rdap+-v+codepen.io vs https://www.whois.com/whois/codepen.io ), but this would be better than nothing and RDAP should be the path for this check going forward.

**Comment by nogweii** [comment]
On May 25, 2023 5:28:53 PM UTC, Nicholas Istre ***@***.***> wrote:
>Looks like not all TLDs have working RDAP yet (`.io` seems to be one, see: https://openrdap.org/demo?cmd=rdap+-v+codepen.io vs https://www.whois.com/whois/codepen.io ), but this would be better than nothing and RDAP should be the path for this check going forward.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/louislam/uptime-kuma/issues/944#issuecomment-1563263456
>You are receiving this because you commented.
>
>Message ID: ***@***.***>
Yeah, all of the generic TLDs must implement RDAP per ICANN policies, but the country codes are free to do whatever their government wants.

**Comment by smitty-exos** [comment]
I hope this gains some tractions and seems like it would be trivial to add for the most part. Yes some of the TLDs aren't doing RDAP, in those cases, a simple whois seems to work just fine. I get it, there is no standardization around whois and what is returned....But I think RDAP would cover a large chunk of domains out there.

**Comment by pavo976** [comment]
You are right, the output of WHOIS is always different, except for the line:

```
Registry Expiry Date:
```
Although, just as mentioned above, I also did not find these details in the [RFC](https://datatracker.ietf.org/doc/html/rfc7485#section-5.2.1)
I haven't seen output for any existing domain name that doesn't have an expiration line, please correct me if the expiration output is different somewhere.

**Comment by zimbres** [comment]
I have a Remote Probe implementation to work with Push monitor type that has a function to check for Domain Expiration. May help somebody.

[Branch ](https://github.com/zimbres/UptimeKumaRemoteProbe/tree/endpoints_sync)

**Comment by Computroniks** [comment]
Looking at some stats about RDAP support (source https://deployment.rdap.org/) it looks like 72.1% of domains support it, with all gTLDs supporting it but only 22.7% of ccTLDs supporting. Given that 72.1% of domains support it, I would be inclined to suggest that we start by implementing a domain expiry using RDAP only. That should simplify implementation a bit hopefully and still cover the majority of use cases. If in future it is determined that accessing those domains that don't currently support RDAP, we could perhaps think about how we want to deal with that problem (maybe if enough people ask ccTLDs for RDAP, they will provide? :wink:). 

I think that I might have a bit of a play with this over the coming days and see if I can get anything working. As far as implementation, would we want to create an entirely new monitor type (something with a long interval), or something more like what we have with the certificate expiry, with the monitors URL / hostname being used as the domain to check?

**Comment by CommanderStorm** [comment]
I would argue that this should be a dedicated monitor, but this is just 
- my feeling of how generally useful (as in useful for everybody),
- technical debt (how complex is the monitor?) and
- UX-friedly (there are plenty of options for monitors already ⇒ adding more (without them being required) might be overwhelming)

If you decide that this is a good decision, please have a look at the new way of writing monitors (#3178) instead of the one used in older PRs ^^

**Comment by mikekosulin** [comment]
[whoiser](https://www.npmjs.com/package/whoiser) package works great.
The initial check should occur when adding a site and subsequent checks should be carried out periodically, similar to how SSL certificate checks are done.

**Comment by fjc0k** [comment]
This might be used as an alternative: #3633 .

**Comment by TheLinuxGuy** [comment]
+1

I have several domain names where I would want to get a 90 day advance expiration (or WHOIS data change notification). This to allow me to find the cheapest domain renewal rates and so on.

**Comment by CommanderStorm** [comment]
@TheLinuxGuy 
Please refrain from posting `+1` / requests for updates things on issues, as this makes issue-management harder.
Issues are for discussing **what needs to be done how by whom**.
[We use 👍🏻 on issues to prioritise work](https://github.com/louislam/uptime-kuma/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc), as always: [Pull Requests](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) welcome.

An example-pr that implements a monitor according to our new syntax (not just piling anything into `monitor.js`) can be found here: https://github.com/louislam/uptime-kuma/pull/3178 or https://github.com/louislam/uptime-kuma/pull/3308

**Comment by mstgelais** [comment]
Wouldn't it possible to use the Linux `whois` command somehow?

**Some output examples**

```
whois google.com
...
Registry Expiry Date: 2028-09-14T04:00:00Z
```

```
whois github.com
...
Registrar Registration Expiration Date: 2024-10-09T00:00:00+0000
```

**Comment by CommanderStorm** [comment]
Note that this is not blocked on a technical level, but on the implementation side.
If you'd like to propose a monitor here is our contribution guide:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md

As noted above, `whois` does not have a stable interface.
⇒ `RDAP` would be preferable for the TLDs which support it

**Comment by thefeli73** [comment]
fetching domain expiration has been implemented for `whois` *and* `RDAP` in the ["domain_exporter" for prometheus by caarlos0](https://github.com/caarlos0/domain_exporter)! 
It's written in go, but perhaps someone could use it as inspiration for implementing it here in js? I lack the nodejs competence to do it myself, but hoping this helps :)

**Comment by CommanderStorm** [comment]
@tahaipek 
No change in comparison to the previous state.
Please refer to my previous comment https://github.com/louislam/uptime-kuma/issues/944#issuecomment-1756306092:

> Note that this is not blocked on a technical level, but on the implementation side.
If you'd like to propose a monitor, here is our contribution guide:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md
> An example-pr that implements a monitor according to our new syntax (not just piling anything into monitor.js) can be found here: https://github.com/louislam/uptime-kuma/pull/3178 or https://github.com/louislam/uptime-kuma/pull/3308
  
> As noted above, whois does not have a stable interface.
⇒ RDAP would be preferable for the TLDs which support it. [node-rdap](https://github.com/dcoffey-zengenti/node-rdap) seems like a solid choice, but we would be open to alternatives if somebody knows a better one ^^

**Comment by MrTomek** [comment]
Hi @louislam 
Do we have any progress in the plan to implement this functionality ?

**Comment by ivancarlosti** [comment]
> Hi @louislam Do we have any progress in the plan to implement this functionality?

I made this using n8n. It's particularly challenging because each whois provides the date/time expiration in a different format, I don't think that will be implemented soon here :/

**Comment by becker** [comment]
Hello!
I think [RDAP](https://www.rfc-editor.org/rfc/rfc7480.txt) could be used to monitor this.

Some API examples:
[OpenRDAP](https://www.openrdap.org/api)
[RDAP.org](https://about.rdap.org/)

**Comment by ivancarlosti** [comment]
> Hi @louislam Do we have any progress in the plan to implement this functionality ?

I made this using n8n. It's particularly challenging because each whois provides the date/time expiration in a different format, I don't think that will be implemented soon here :/

**Comment by ivancarlosti** [comment]
> Looking at some stats about RDAP support (source https://deployment.rdap.org/) it looks like 72.1% of domains support it, with all gTLDs supporting it but only 22.7% of ccTLDs supporting. Given that 72.1% of domains support it, I would be inclined to suggest that we start by implementing a domain expiry using RDAP only. That should simplify implementation a bit hopefully and still cover the majority of use cases. If in future it is determined that accessing those domains that don't currently support RDAP, we could perhaps think about how we want to deal with that problem (maybe if enough people ask ccTLDs for RDAP, they will provide? 😉).
> 
> I think that I might have a bit of a play with this over the coming days and see if I can get anything working. As far as implementation, would we want to create an entirely new monitor type (something with a long interval), or something more like what we have with the certificate expiry, with the monitors URL / hostname being used as the domain to check?
> 
> I know about #1775 but that make use of whois which we know to be a bit problematic and it hasn't been updated for over a year.

It's not true that all gTLD support it, .ES of Spain for example do not have it.

**Comment by lfalcao** [comment]
Hello, there's any progress here? anyone is working or have done anything here?

**Comment by enter1010** [comment]
+1

**Comment by xardit** [comment]
Using a WHOIS CLI directly or a JSON API with whoisjson.com seems like the right way. Adding a Docker container to provide that service might be a solution, though it's a bit complex.

**Comment by CommanderStorm** [comment]
@xardit as indicated above, using whois for this is not the best approach. 
RDAP fixes this and rollout is currently most country's and all non-country TLDs.

Adding a system dependency for this or external API does not seem nessedary.

I can review PRs, but don't currently have the time to invest into adding monitors.

**Comment by mrjones-plip** [comment]
While not at all a native feature to Uptime Kuma, I recently wrote an app that takes a list of domains and caches their expiration info in JSON files.  These JSON files are then expose via an HTTP sever so Uptime Kuma can easily monitor and alert on them via the `HTTP(s) Json Query` feature.  See [the repo for details](https://github.com/mrjones-plip/GEA?tab=readme-ov-file#uptime-kuma)!

**Comment by starnetwork** [comment]
this can be very good as additional Advanced option to all HTTP(s) monitors
Technically this is possible, similar to expiration of the SSL certificate option.

**Comment by mabed-fr** [comment]
Yes is a good idea

**Comment by CommanderStorm** [comment]
Feel free to test the following PR:
- https://github.com/louislam/uptime-kuma/pull/6098

Here is our section on the wiki how to do this:
https://github.com/louislam/uptime-kuma/wiki/Test-Pull-Requests

**Comment by CommanderStorm** [comment]
Here is the new PR:
- https://github.com/louislam/uptime-kuma/pull/6413

It needs some testing, code wise it looks good as said above

**Comment by beztebya666** [comment]
You can do it through flow:

check RDAP(success?) -> check WHOIS(success?) -> ERROR

**Comment by CommanderStorm** [comment]
We now have something that mostly works imo:
- https://github.com/louislam/uptime-kuma/pull/6413

If there are enhancements that can be made, lets put them into separate issues ^^

**Comment by imranrazakhan** [comment]
@CommanderStorm  I have deployed uptime-kuma on Kubernetes using Helm, and it works correctly up to version ` louislam/uptime-kuma:2.1.0-beta.0`

After upgrading to a newer version `uptime-kuma:2.1.0-beta.1`, the deployment no longer works and the pod starts throwing the following error:

```
2026-01-13T14:51:00+01:00 [RDAP] DEBUG: FetchError: request to https://data.iana.org/rdap/dns.json failed, reason:
    at ClientRequest.<anonymous> (file:///app/node_modules/node-fetch-cache/node_modules/node-fetch/src/index.js:108:11)
    at ClientRequest.emit (node:events:519:28)
    at emitErrorEvent (node:_http_client:107:11)
    at TLSSocket.socketErrorListener (node:_http_client:574:5)
    at TLSSocket.emit (node:events:519:28)
    at emitErrorNT (node:internal/streams/destroy:170:8)
    at emitErrorCloseNT (node:internal/streams/destroy:129:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:90:21) {
  type: 'system',
  errno: 'ETIMEDOUT',
  code: 'ETIMEDOUT',
  erroredSysCall: undefined
}
2026-01-13T14:51:00+01:00 [DOMAIN_EXPIRY] WARN: Domain expiry unsupported for '.com' because its RDAP endpoint is not listed in the IANA database.
2026-01-13T14:51:02+01:00 [SETTINGS] DEBUG: Get Setting (cache): trustProxy: null
2026-01-13T14:51:02+01:00 [ENTRY] DEBUG: Request Domain: 127.0.0.1
2026-01-13T14:51:05+01:00 [SETTINGS] DEBUG: Get Setting (cache): trustProxy: null
2026-01-13T14:51:05+01:00 [ENTRY] DEBUG: Request Domain: 100.67.131.246
2026-01-13T14:51:12+01:00 [SETTINGS] DEBUG: Get Setting (cache): trustProxy: null
2026-01-13T14:51:12+01:00 [ENTRY] DEBUG: Request Domain: 127.0.0.1
2026-01-13T14:51:15+01:00 [SETTINGS] DEBUG: Get Setting (cache): trustProxy: null
2026-01-13T14:51:15+01:00 [ENTRY] DEBUG: Request Domain: 100.67.131.246
2026-01-13T14:51:22+01:00 [SETTINGS] DEBUG: Get Setting (cache): trustProxy: null
2026-01-13T14:51:22+01:00 [ENTRY] DEBUG: Request Domain: 127.0.0.1
2026-01-13T14:51:25+01:00 [SETTINGS] DEBUG: Get Setting (cache): trustProxy: null
2026-01-13T14:51:25+01:00 [ENTRY] DEBUG: Request Domain: 100.67.131.246
2026-01-13T14:51:28+01:00 [SETTINGS] DEBUG: Cache Cleaner is just started.
```

The issue appears only after upgrading past beta-0; no other configuration changes were made.

Please let me know if any additional information  is required.

**Comment by CommanderStorm** [comment]
Could you open an issue for all remaining issues?
I can't work with comments on resolved issues as I don't have the time to fix all issues.

About this issue:
Hmm.. not sure how to handle this case properly.
Are there any network restrictions between the pod and 

Since the list is fairly static, likely we should just add this to the repo instead to avoid this failure.
